src/server/page_server.cpp

```diff
@@ -205,6 +206,13 @@ page_server::connection_handler::receive_start_catch_up (tran_server_conn_t::seq
 
   er_log_debug (ARG_FILE_LINE, "receive_start_catch_up: hostname = %s, port = %d, LSA = (%lld|%d)\n", host.c_str (), port,
 		LSA_AS_ARGS (&catchup_lsa));
+  if (port == -1)
+    {
+      return; // TODO: It means that the ATS is booting up, it will be set properly after ATS recovery is implemented.
+    }
+
+  // TODO: A thread will take the catch-up including establishing connection to avoid blocking ATS->PS reqeusts.
+  m_ps.connect_to_leader_page_server (std::move (host), port);
```

Comment:
@wrlawodms
Remark.
Is it safe to assume that `page_server::connection_handler` class will also need to implement some state transition similar to `tran_server::connection_handler`?
Conceptual states:
1. `IDLE`
2. `CONNECTING` (ATS/PTS connecting to PS, maybe not needed)
3. `CATCHUP_CONNECTING` (PS needs to perfom catch-up and is connecting to the leader PS)
4. `CATCHUP_CATCHING_UP` (PS connected to leader PS and performing catching up)
5. `CATCHUP_DISCONNECTING` (PS done catching up and is disconnecting from leader PS, maybe not needed, jump straight to next one)
6. `CONNECTED` (PS online; either catch-up has been done or not needed)
7. `DISCONNECTING` (ATS/PTS disconnecting from PS, maybe not needed)

Transitions:
- 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> 7 -> 1
- 1 -> 2 -> 6 -> 7 -> 1

Reply:
Thank you for thinking thoroughly. In my opinion, we can define the state transition, but I can't figure out its necessity yet. If there are some operations only allowed in some states, I think that's the time to introduce the states. The `page_esrver::conneciton_handler` is even a transient object while the TS's is permanent. What do you think? Do you think we'd be better off introducing the states in advance?

Reply:
No, I don't think it is needed right now. As you say, let's keep it simple for now.

```
The page_esrver::conneciton_handler is even a transient object while the TS's is permanent.
```
I disagree with this.

I see the following structure moving forward - which is very much in line and a continuation of what is added by this PR:
- the one flavor currently existing `page_server::connection_handler` which is the passive one:
  - serves requests from xTSs
    - there is currently one instance per connected xTs
    - and this is very much permanent, not transient
  - this connection will be extended to respond to requests from the "follower" PS
    - no need to have yet another implementation on the "leader" side
    - for the connected follower PS a new - transient - instance of such a connector will be added by the leader PS and removed once the follower PS will disconnect (having performed the catch-up)
- a new flavor of connection handler - `catchup_connection_handler` which will be instantiated by the follower PS and will live as long as there is catchup to do
  - this new flavor is the one that is instantiated at about where we have now the line:
```
  // TODO: A thread will take the catch-up including establishing connection to avoid blocking ATS->PS reqeusts.
  m_ps.connect_to_leader_page_server (std::move (host), port);
```
  - and I see here a need for the states being introduced - the `page_server` object instance needs to "play" with this - conceptual entity - "catchup_connection_handler" asynchronously
  - the conceptual state of the `page_server` instance is suspended as long as the `catchup_connection_handler` has not finished it work in one way (success) or another (failure).

This is just a suggestion of how the structure might evolve. It may very well be that you have another - maybe clearer - picture.

Reply:
In the light of what I described above, my first comment about the states and transitions is .. well .. out of date/incorrect/incomplete.

Reply:
We have a different picture for the connections. Lets me share mine, which is described other issues, but I understand it's hard to get together.

A page server will have two connection_handlers
- tran_server_connection_handler
  - it's only used to communicate with a transaction server.
- catchup_connection_handler
  - through which PSs communicate with each other. Both a follower and a leader use this connection handler. This will have both the request and its handler for the catch-up.

Let me depict them:
```

                 PS1                                                                                                      
     +------------------------------+                                                                                      
     |tran_server_connection_handler   <-->   connection_handler                 
   --|catch_upconnection_handler    |                                                                                      
   | +------------------------------+                                                                                      
   | |tran_server_connection_handler|  <-->   connection_handler                                                           
   +-|catchup_connection_handler    |                                                                                      
     +------------------------------+                                                                                      
                 PS2           
```                   

I can't imagine how we can handler requests of both TS and PS in one connection_handler. The internal server object, `request_sync_client_server`, define requests it can send and receive like `page_to_tran_request` or `tran_to_page_request`. We can add two `tran_to_page_request`, but I think it would get complicated. My plan is to define a new request type, `page_to_page_request` and create a connection_handler using it, which will be totally separate from the connection_handler with a TS.

And I think a page_server can communicate with a tran_server regardless of the state of its catchup_connection_handler.

Reply:
> The page_server::connection_handler on a leader PS already has implemented the handler to server log pages - tran_to_page_request::SEND_LOG_PAGE_FETCH. It will probably need to implement some other handlers for extra requests as well.
So the existing connection_handler will be extended to stay on the "serving" side - from the point of view of the PS-PS catchup mechanism.

We need two internal `request_sync_client_server` in the page_server::connection_handler.
- cubcomm::request_sync_client_server<page_to_tran_request, tran_to_page_request, std::string>;
- cubcomm::request_sync_client_server<page_to_page_request, page_to_page_request, std::string>;

Do you mean this? Or, we can change tran_to_page_request -> xxx_to_page_request and combine the catchup_requests and tran_to_page_request, but it seems better to me to split them.

Besides, I guess we can simplify the serving side connection handler for the catchup by implementing only what's needed. I'll consider your approach too when implementing the connection_handler in the following issue. Let's see in the issue or PR.